libtiny_client/src/state.rs

```diff
@@ -207,21 +278,23 @@ impl StateInner {
                         self.usermask = Some(usermask);
 
                         // Initialize channel state
-                        match utils::find_idx(&self.chans, |(s, _)| s == chan) {
+                        match utils::find_idx(&self.chans, |c| &c.name == chan) {
                             None => {
-                                self.chans.push((chan.to_owned(), HashSet::new()));
+                                self.chans.push(Chan::new(chan.to_owned()));
                             }
                             Some(chan_idx) => {
                                 // This happens because we initialize channel states for channels
                                 // that we will join on connection when the client is first created
-                                self.chans[chan_idx].1.clear();
+                                let chan = &mut self.chans[chan_idx];
+                                chan.join_state = JoinState::Joined;
```

Comment:
This is the point where the JoinState gets set to Joined -- once the channel is actually joined